v3/pattern.go

```diff
@@ -1,34 +1,70 @@
 package v3
 
-// The pattern leaf and pattern tree sentinels allow definining patterns in
-// terms of the expression structure. A pattern expression is used to extract
-// an actual expression from the memo in order to perform a transformation on
-// it. (See memo.{bind,advance}). The pattern sentinels indicate where
-// recursive extraction of the full subtree is required. For example, the join
-// commutativity transformation wants to extract expressions. It does not care
-// about the inputs, but only wants to reorder them. This can be specified by
-// the pattern expression:
+// A pattern expression is used to extract expressions from the memo in order to
+// perform a transformation on them. The pattern expression describes what the
+// structure of the expression needs to be for a given transformation to apply.
+// A memo expression can have many bindings that satisfy a given pattern.
 //
-// inner join
-//   0: pattern leaf
-//   1: pattern leaf
-//   2: pattern leaf
+// To help define these patterns, we define pattern tree sentinels which can be
+// used as leaves of the pattern expression tree.
 //
-// Join associativity is somewhat more complicated in that it wants a join
-// where the left input is also a join:
+//  1. Pattern Leaf
 //
-// inner join
-//   0: inner join
-//     0: pattern leaf
-//     1: pattern leaf
-//     2: pattern tree
-//   1: pattern leaf
-//   2: pattern tree
+//  A Pattern Leaf matches any expression tree, with only the root of the tree
+//  being retained in a binding. It is used when the expression is used opaquely
+//  by the transformation. In other words, the transformation doesn't care
+//  what's inside the subtree. It is a "leaf" in the sense that it's a leaf in
+//  any binding matching a pattern.
+//
+//  1. Pattern Tree
```

Comment:
`s/1/2/g`. Or perhaps get rid of the numbers.

Reply:
Removed the numbers